cardano-ping/src/Cardano/Network/Ping.hs

```diff
@@ -381,24 +434,59 @@ pingClient stdout stderr PingOpts{pingOptsQuiet, pingOptsJson, pingOptsCount} ve
     let timeout = 30
     bearer <- getBearer makeSocketBearer timeout nullTracer sd
 
-    !t1_s <- write bearer timeoutfn $ wrap handshakeNum InitiatorDir (handshakeReq versions)
+    !t1_s <- write bearer timeoutfn $ wrap handshakeNum InitiatorDir (handshakeReq versions pingOptsHandshakeQuery)
     (msg, !t1_e) <- nextMsg bearer timeoutfn handshakeNum
     unless pingOptsQuiet $ printf "%s handshake rtt: %s\n" peerStr (show $ diffTime t1_e t1_s)
 
     case CBOR.deserialiseFromBytes handshakeDec msg of
       Left err -> eprint $ printf "%s Decoding error %s" peerStr (show err)
       Right (_, Left err) -> eprint $ printf "%s Protocol error %s" peerStr (show err)
-      Right (_, Right version) -> do
-        unless pingOptsQuiet $ printf "%s Negotiated version %s\n" peerStr (show version)
-        keepAlive bearer timeoutfn peerStr version (tdigest []) 0
-        -- send terminating message
-        _ <- write bearer timeoutfn $ wrap keepaliveNum InitiatorDir (keepAliveDone version)
-        -- protocol idle timeout
-        MT.threadDelay 5
+      Right (_, Right recVersions) -> do
+        when (pingOptsHandshakeQuery && not pingOptsQuiet) $ printf "%s Queried versions %s\n" peerStr (show recVersions)
+        case acceptVersions recVersions of
+          Left err ->
+            eprint $ printf "%s Version negotiation error %s\n" peerStr err
+          Right version -> do
+            unless pingOptsQuiet $ printf "%s Negotiated version %s\n" peerStr (show version)
+            unless pingOptsHandshakeQuery $
+              keepAlive bearer timeoutfn peerStr version (tdigest []) 0
+            -- send terminating message
+            _ <- write bearer timeoutfn $ wrap keepaliveNum InitiatorDir (keepAliveDone version)
```

Comment:
```suggestion
            unless pingOptsHandshakeQuery $
              keepAlive bearer timeoutfn peerStr version (tdigest []) 0
              -- send terminating message
              _ <- write bearer timeoutfn $ wrap keepaliveNum InitiatorDir (keepAliveDone version)
              return ()
```

Reply:
The `keepAlive` termination message also shouldn't be send.  

It ought to be part of the `keepAlive` function.